cardano-cli/src/Cardano/CLI/Ping.hs

```diff
@@ -0,0 +1,582 @@
+    unless pingCmdQuiet $ printf "%s handshake rtt: %s\n" peerStr (show $ diffTime t1_e t1_s)
+
+    case CBOR.deserialiseFromBytes handshakeDec msg of
+      Left err -> eprint $ printf "%s Decoding error %s\n" peerStr (show err)
+      Right (_, Left err) -> eprint $ printf "%s Protocol error %s\n" peerStr (show err)
+      Right (_, Right version) -> do
+        unless pingCmdQuiet $ printf "%s Negotiated version %s\n" peerStr (show version)
+        keepAlive bearer timeoutfn peerStr version (tdigest []) 0
+        -- send terminating message
+        _ <- write bearer timeoutfn $ wrap keepaliveNum InitiatorDir (keepAliveDone version)
+        -- protocol idle timeout
+        threadDelay 5
+
+  )
+  where
+
+    peerString :: IO String
+    peerString =
+      case Socket.addrFamily peer of
+        Socket.AF_UNIX -> return $ show $ Socket.addrAddress peer
+        _ -> do
+          (Just host, Just port) <-
+            Socket.getNameInfo
+              [Socket.NI_NUMERICHOST, Socket.NI_NUMERICSERV]
+              True True (Socket.addrAddress peer)
+          return $ host <> ":" <> port
+
+    toSample :: Time -> Time -> Double
+    toSample t_e t_s = realToFrac $ diffTime t_e t_s
+
+    eprint :: String -> IO ()
+    eprint = BS.Char.hPutStr stderr . BS.Char.pack
+
+    nextMsg ::  MuxBearer IO -> TimeoutFn IO -> MiniProtocolNum -> IO (BL.ByteString, Time)
+    nextMsg bearer timeoutfn ptclNum = do
+      (sdu, t_e) <- Network.Mux.Types.read bearer timeoutfn
+      if mhNum (msHeader sdu) == ptclNum
+        then return (msBlob sdu, t_e)
+        else nextMsg bearer timeoutfn ptclNum
+
+    keepAlive :: MuxBearer IO
+              -> TimeoutFn IO
+              -> String
+              -> NodeVersion
+              -> TDigest 5
+              -> Word32
+              -> IO ()
+    keepAlive _ _ _ _ _ cookie | cookie == pingCmdCount = return ()
+    keepAlive bearer timeoutfn peerStr version td !cookie = do
+      let cookie16 = fromIntegral cookie
+      !t_s <- write bearer timeoutfn $ wrap keepaliveNum InitiatorDir (keepAliveReq version cookie16)
+      (!msg, !t_e) <- nextMsg bearer timeoutfn keepaliveNum
+      let rtt = toSample t_e t_s
+          td' = insert rtt td
+      case CBOR.deserialiseFromBytes (keepAliveRspDec version) msg of
+        Left err -> eprint $ printf "%s keepalive decoding error %s\n" peerStr (show err)
+        Right (_, Left err) -> eprint $ printf "%s keepalive protocol error %s\n" peerStr (show err)
+        Right (_, Right cookie') -> do
+          when (cookie' /= cookie16) $ eprint $ printf "%s cookie missmatch %d /= %d\n"
+            peerStr cookie' cookie
+
+          now <- getCurrentTime
+          let point = toStatPoint now peerStr cookie16 rtt td'
+          if pingCmdJson
+            then traceWith tracer $ LogMsg (encode point)
+            else traceWith tracer $ LogMsg $ BL.Char.pack $ show point <> "\n"
+          hFlush stdout
+          threadDelay 1
```

Comment:
I'm assuming you don't want to delay for a single microsecond and just want to allow a context switch
```suggestion
          yield
```
If you actually want to sleep, you probably want a much longer duration than a microsecond

Reply:
Never mind, that `threadDelay` comes from `ouroboros-network` and takes a `DiffTime`, not an `Int` like the `threadDelay` from `Control.Concurrent`.
Personally, I find that slightly confusing. Given that `Control.Concurrent.threadDelay 1` and `Control.Monad.Class.Timers.threadDelay 1` both type check here and have different results I'm slightly uncomfortable. Can we require uses of `threadDelay` on literals to either be qualified, or to explicitly give the literal a type? Otherwise we might introduce bugs when refactoring and someone's "helpful" IDE imports the wrong `threadDelay`.

Reply:
I made `threadDelay` a qualified import.

Reply:
I made `threadDelay` a qualified import.